### PR TITLE
Refactor SDLPrinter for better indentation

### LINKED ISSUE #287: Strange formatting of input types with descriptions
If an input type has field descriptions then formatting produces unexpected extra new lines and extra indentation of fields.

Please find attached the project for reproducing the issue.
[TryGraphQLParser.zip](https://github.com/graphql-dotnet/parser/files/10471734/TryGraphQLParser.zip)

**Test1**

No issues for `type` types:

    dotnet run -- x1-type.graphql

- input `x1-type.graphql` contains a `type` with fields with descriptions.
- output `x1-type.graphql.output.graphql` is formatted as expected.

**Test2**

Formatting issues with `input` with descriptions:

    dotnet run -- x2-input.graphql

- input `x2-input.graphql` contains an `input` with fields with descriptions.
- output `x2-input.graphql.output.graphql` shows unexpected new lines and extra indentations.

**Test3**

No issues with `input` without descriptions:

    dotnet run -- x3-input.graphql

- input `x3-input.graphql` contains an `input` with no descriptions.
- output `x3-input.graphql.output.graphql` is formatted as expected.

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #287
waits #303 to be merged into 

The core idea of this PR - manage indentation more centrally. Side note - I tried to get rid of trailing NewLine when printing AST.

### CODE REVIEW

File: src/GraphQLParser.ApiTests/GraphQLParser.approved.txt [hunk 1]
```diff
@@ -852,7 +856,11 @@ namespace GraphQLParser.Visitors
     public interface IPrintContext : GraphQLParser.Visitors.IASTVisitorContext
     {
         int IndentLevel { get; set; }
+        bool IndentPrinted { get; set; }
```
Comment (sungam3r) [inline src/GraphQLParser.ApiTests/GraphQLParser.approved.txt:859]: @Shane32 Formally, this is breaking changes, but I doubt that there are a sufficient number of people for whom it will be a real problem

File: src/GraphQLParser.ApiTests/GraphQLParser.approved.txt [hunk 2]
```diff
@@ -895,9 +907,9 @@ namespace GraphQLParser.Visitors
     public class SDLPrinterOptions
     {
         public SDLPrinterOptions() { }
-        public int IndentSize { get; set; }
         public bool EachDirectiveLocationOnNewLine { get; init; }
         public bool EachUnionMemberOnNewLine { get; init; }
+        public int IndentSize { get; init; }
```
Comment (sungam3r) [inline src/GraphQLParser.ApiTests/GraphQLParser.approved.txt:912]: To align with other modifiers. Changes from #294 were not published yet.

File: src/GraphQLParser.Tests/Visitors/SDLPrinterTests.cs [hunk 3]
```diff
@@ -697,15 +764,13 @@ public async Task SelectionSet_Without_Parent_Should_Be_Printed_On_New_Line()
         var printer = new SDLPrinter(new SDLPrinterOptions { PrintComments = true });
         await printer.PrintAsync(selectionSet, writer);
         writer.ToString().ShouldBe(@"{
-}
-");
+}");
```
Comment (sungam3r) [inline src/GraphQLParser.Tests/Visitors/SDLPrinterTests.cs:767]: Better handling for NewLine at the end of printed AST.

File: src/GraphQLParser.ApiTests/GraphQLParser.approved.txt [hunk 4]
```diff
@@ -906,6 +918,7 @@ namespace GraphQLParser.Visitors
         public SDLPrinter() { }
         public SDLPrinter(GraphQLParser.Visitors.SDLPrinterOptions options) { }
         public GraphQLParser.Visitors.SDLPrinterOptions Options { get; }
+        protected virtual System.Threading.Tasks.ValueTask MakeVerticalIndentationBetweenTopLevelDefinitions(GraphQLParser.AST.ASTNode node, TContext context) { }
```
Comment (sungam3r) [inline src/GraphQLParser.ApiTests/GraphQLParser.approved.txt:921]: New method is needed for descendants that may skip nodes, see `PrintOnlyStartsWithA`. Without it `SDLPrinter` prints large vertical indentation in place of skipped nodes.

File: src/GraphQLParser.Tests/Visitors/SDLPrinterTests.cs [hunk 5]
```diff
@@ -575,6 +575,73 @@ type Query
      subscription: S
 }
 """, true, false, false, 5)]
+    [InlineData(45,
+"""
+"A component contains the parametric details of a PCB part."
```
Comment (sungam3r) [inline src/GraphQLParser.Tests/Visitors/SDLPrinterTests.cs:580]: @nightroman your test now works

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs [hunk 6]
```diff
@@ -15,27 +15,34 @@ public static class PrintContextExtensions
...
     /// </summary>
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static ValueTask WriteAsync<TContext>(this TContext context, ROM value)
+    public static async ValueTask WriteAsync<TContext>(this TContext context, ROM value)
         where TContext : IPrintContext
     {
+        if (value.Length == 0)
+            return;
+
+        context.NewLinePrinted = value.Span[value.Length - 1] == '\n';
```
Comment (sungam3r) [inline src/GraphQLParser/Visitors/PrintContextExtensions.cs:24]: That approach is much easier than trying to calculate the flag in each place where `WriteAsync` is called.

**Comment by sungam3r** [comment]
I'm sure there are more issues to fix but for now I finished. I simplified working with indents, especially for comments and descriptions and added more tests to verify edge cases.

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs [hunk 7]
```diff
@@ -15,27 +15,34 @@ public static class PrintContextExtensions
     /// using <see cref="CancellationToken"/> from it.
     /// </summary>
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static ValueTask WriteAsync<TContext>(this TContext context, ROM value)
+    public static async ValueTask WriteAsync<TContext>(this TContext context, ROM value)
         where TContext : IPrintContext
     {
+        if (value.Length == 0)
+            return;
```
Comment (Shane32) [inline src/GraphQLParser/Visitors/PrintContextExtensions.cs:22]: I suggest `return default;` and remove `async` and remove `await task`.  Whatever you want though.

**Review by Shane32 — APPROVED** [review]

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs (hunk 7 above)
Comment (sungam3r) [reply src/GraphQLParser/Visitors/PrintContextExtensions.cs:22]: Ah, I forgot to revert it after some of my previous changes.

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs [hunk 8]
```diff
@@ -15,27 +15,34 @@ public static class PrintContextExtensions
     /// using <see cref="CancellationToken"/> from it.
     /// </summary>
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public static ValueTask WriteAsync<TContext>(this TContext context, ROM value)
+    public static async ValueTask WriteAsync<TContext>(this TContext context, ROM value)
```
Comment (sungam3r) [inline src/GraphQLParser/Visitors/PrintContextExtensions.cs:18]:
```suggestion
    public static ValueTask WriteAsync<TContext>(this TContext context, ROM value)
```

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs (hunk 7 above)
Comment (sungam3r) [inline src/GraphQLParser/Visitors/PrintContextExtensions.cs:22]:
```suggestion
            return default;
```

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs [hunk 9]
```diff
@@ -15,27 +15,34 @@ public static class PrintContextExtensions
...
         var task =
 #if NETSTANDARD2_0
         // no cancellationToken support on netstandard2.0
         context.Writer.WriteAsync(value.ToString()); //ISSUE: allocation - either WriteAsync(value.ToString()) or Write(char value) in a loop
 #elif NETSTANDARD2_1_OR_GREATER || NET6_0_OR_GREATER
-            context.Writer.WriteAsync(value, context.CancellationToken);
+        context.Writer.WriteAsync(value, context.CancellationToken);
 #endif
-        return new ValueTask(task);
+        await task;
```
Comment (sungam3r) [inline src/GraphQLParser/Visitors/PrintContextExtensions.cs:34]:
```suggestion
        return new ValueTask(task);
```

File: src/GraphQLParser/Visitors/PrintContextExtensions.cs (hunk 7 above)
Comment (sungam3r) [reply src/GraphQLParser/Visitors/PrintContextExtensions.cs]: Done

**Comment by sungam3r** [comment]
So what about version? 8.3 or 9.0? I lean to 8.3 since printer is auxiliary code and not too many people should use custom contexts for printing.

**Comment by sungam3r** [comment]
More argument against 9.0 - it will require to bump version of core report and server.

**Comment by Shane32** [comment]
Your call. If it’s source compatible then I’m okay with 8.3. 

The core repo and server specify 8.2+ probably and don’t need to bump to 9.0 if they don’t rely on any breaking API.

**Comment by Shane32** [comment]
Whatever you think I guess. It really only matters to you if the features are not used by GraphQL. I’d just publish as 9.0 and leave GraphQL 7.x using parser 8.x until we bump GraphQL to 8.0. Again your call.

**Comment by sungam3r** [comment]
It is not source compatible in case of custom printer context implemented in user code. I added new members in public interface.

**Comment by sungam3r** [comment]
I have never used that feature before, will look tomorrow.

**Comment by sungam3r** [comment]
Instance state is not allowed for default interfaces so I'll bump to 9.
